lib/Sema/CSApply.cpp

```diff
@@ -2514,6 +2514,33 @@ namespace {
             /*implicit=*/expr->isImplicit(), Type(), getType);
         result = finishApply(apply, Type(), cs.getConstraintLocator(expr));
       }
+			
+      // If the base type is optional and we're referring to an enum case via the dot syntax
+      // and the enum case matches the Optional<T>.none case then emit a diagnostic
+      if (auto baseType = baseTy->lookThroughAllOptionalTypes()) {
+        if (auto DSCE = dyn_cast<DotSyntaxCallExpr>(result)) {
+          if (auto EED = dyn_cast<EnumElementDecl>(DSCE->getCalledValue())) {
+            if (EED->getParentEnum()->isOptionalDecl()) {
+              // Lookup the case in the original enum decl
+              auto baseTypeEnum = baseType->getEnumOrBoundGenericEnum();
+              auto results = tc.lookupMember(baseTypeEnum->getModuleContext(), baseType, EED->getName(), defaultMemberLookupOptions);
+              bool caseExistsInBase = false;
+              
+              for (LookupResultEntry result : results) {
+                if (auto *eed = dyn_cast<EnumElementDecl>(result.getValueDecl())) {
+                  caseExistsInBase = true;
```

Comment:
Question: Do you want to emit an error if `case none` has an associated value? I'm not sure if `.none(value)` would select the base type's case instead of `Optional`'s—if it would, you might want to suppress the warning.